code/libs/ffmpeg/FFmpegContext.cpp

```diff
@@ -78,61 +135,76 @@ FFmpegContext::~FFmpegContext() {
 
-	if (!instance->m_ctx) {
-		throw FFmpegException("Failed to allocate context!");
-	}
+    if (!avioBuffer) {
+        throw FFmpegException("Failed to allocate IO buffer!");
+    }
 
-	auto avioBuffer = reinterpret_cast<uint8_t*>(av_malloc(AVIO_BUFFER_SIZE));
+    auto ioContext =
+        avio_alloc_context(avioBuffer, AVIO_BUFFER_SIZE, 0, opaque, read_packet, nullptr, seek);
 
-	if (!avioBuffer) {
-		throw FFmpegException("Failed to allocate IO buffer!");
-	}
+    if (!ioContext) {
+        throw FFmpegException("Failed to allocate IO context!");
+    }
 
-	auto ioContext =
-		avio_alloc_context(avioBuffer, AVIO_BUFFER_SIZE, 0, instance->m_file, cfileRead, nullptr, cfileSeek);
+    instance->m_ctx->pb = ioContext;
 
-	if (!ioContext) {
-		throw FFmpegException("Failed to allocate IO context!");
-	}
+    auto probe_ret = av_probe_input_buffer2(instance->m_ctx->pb, &instance->m_ctx->iformat, nullptr, nullptr, 0, 0);
+    if (probe_ret < 0) {
+        char errorStr[1024];
+        av_strerror(probe_ret, errorStr, 1024);
 
-	instance->m_ctx->pb = ioContext;
+        throw FFmpegException(SCP_string("Could not open movie file! Error: ") + errorStr);
+    }
 
-	auto probe_ret = av_probe_input_buffer2(instance->m_ctx->pb, &instance->m_ctx->iformat, nullptr, nullptr, 0, 0);
-	if (probe_ret < 0) {
-		char errorStr[1024];
-		av_strerror(probe_ret, errorStr, 1024);
+    instance->m_ctx->flags |= AVFMT_FLAG_CUSTOM_IO;
 
-		throw FFmpegException(SCP_string("Could not open movie file! Error: ") + errorStr);
-	}
+    auto ret = avformat_open_input(&instance->m_ctx, nullptr, instance->m_ctx->iformat, nullptr);
+    if (ret < 0) {
+        char errorStr[1024];
+        av_strerror(ret, errorStr, 1024);
 
-	instance->m_ctx->flags |= AVFMT_FLAG_CUSTOM_IO;
+        throw FFmpegException(SCP_string("Could not open movie file! Error: ") + errorStr);
+    }
 
-	auto ret = avformat_open_input(&instance->m_ctx, nullptr, instance->m_ctx->iformat, nullptr);
-	if (ret < 0) {
-		char errorStr[1024];
-		av_strerror(ret, errorStr, 1024);
+    ret = avformat_find_stream_info(instance->m_ctx, nullptr);
+    if (ret < 0) {
+        char errorStr[1024];
+        av_strerror(ret, errorStr, 1024);
 
-		throw FFmpegException(SCP_string("Could not open movie file! Error: ") + errorStr);
-	}
+        throw FFmpegException(SCP_string("Failed to get stream information! Error: ") + errorStr);
+    }
 
-	ret = avformat_find_stream_info(instance->m_ctx, nullptr);
-	if (ret < 0) {
-		char errorStr[1024];
-		av_strerror(ret, errorStr, 1024);
+    return instance;
+}
 
-		throw FFmpegException(SCP_string("Failed to get stream information! Error: ") + errorStr);
-	}
+std::unique_ptr<FFmpegContext> FFmpegContext::createContext(CFILE* mediaFile) {
+	Assertion(mediaFile != nullptr, "File pointer must be valid!");
+
+	std::unique_ptr<FFmpegContext> instance(new FFmpegContext(mediaFile));
+
+    return prepareInstance(std::move(instance), instance->m_file, cfileRead, cfileSeek);
```

Comment:
Moving the pointer into the function suggests that ownership is transferred but the value is just returned again which is a bit misleading. Instead, you could also make `prepareInstance` just take a raw pointer and keep the unique pointer in this function.

Reply:
Actually, looking at this again, `instance` shouldn't be passed at all. I'm making `prepareInstance` a nonstatic member, and call it _on_ `instance`